finetune/util/huggingface_interface.py

```diff
@@ -0,0 +1,200 @@
+        FINETUNE_BASE_FOLDER, "model", "huggingface", weights_file
+    )
+    raw_weights_path = os.path.join(
+        FINETUNE_BASE_FOLDER, "model", "huggingface", "raw_" + weights_file
+    )
+    if not os.path.exists(raw_weights_path):
+        tqdl.download(weights_url, raw_weights_path)
+    jl.dump(
+        load_weights_from_hdf5_group_by_name(raw_weights_path, weights_replacement),
+        weights_path,
+    )
+
+    def finetune_featurizer(X, encoder, config, train=False, reuse=None, **kwargs):
+        initial_shape = tf.shape(input=X)
+        X = tf.reshape(X, shape=tf.concat(([-1], initial_shape[-1:]), 0))
+        X.set_shape([None, None])
+        delimiters = tf.cast(tf.equal(X, encoder.delimiter_token), tf.int32)
+
+        token_type_ids = tf.cumsum(delimiters, exclusive=True, axis=1)
+
+        seq_length = tf.shape(input=delimiters)[1]
+
+        eos_idx = tf.argmax(
+            input=tf.cast(delimiters, tf.float32)
+            * tf.expand_dims(
+                tf.range(tf.cast(seq_length, tf.float32), dtype=tf.float32), 0
+            ),
+            axis=1,
+        )
+
+        lengths = lengths_from_eos_idx(eos_idx=eos_idx, max_length=seq_length)
+        mask = tf.sequence_mask(lengths, maxlen=seq_length, dtype=tf.float32)
+
+        with tf.compat.v1.variable_scope("model/featurizer", reuse=reuse):
+            hf_model = hf_featurizer(hf_config_instance)
+
+            if config.low_memory_mode and train:
+                for layer in hf_model.encoder.layer:
+                    layer.__call__ == recompute_grad(layer.__call__, train_vars=layer.trainable_weights)
+
+            kwargs = {
+                "attention_mask": mask,
+                "token_type_ids": token_type_ids,
+                "inputs_embeds": None,
+                "training": train,
+            }
+            call_args = tf_inspect.getargspec(hf_model).args
+            kwargs = {k: v for k, v in kwargs.items() if k in call_args}
+
+            model_out = hf_model(X, **kwargs)
+            if isinstance(model_out, tuple) and len(model_out) > 1:
+                sequence_out, pooled_out, *_ = model_out
+            else:
+                sequence_out = model_out[0]
+                pooled_out = sequence_out[:, 0, :]
+
+            n_embed = pooled_out.shape[-1]
+
+            embed_weights = hf_model.embeddings.word_embeddings
+            features = tf.reshape(
+                pooled_out, shape=tf.concat((initial_shape[:-1], [n_embed]), 0),
+            )
+            sequence_features = tf.reshape(
+                sequence_out, shape=tf.concat((initial_shape, [n_embed]), 0),
+            )
+
+            output_state = {
+                "embed_weights": embed_weights,
+                "features": features,
+                "sequence_features": sequence_features,
+                "lengths": lengths,
+                "eos_idx": eos_idx,
+            }
+
+            return output_state
+
+    class HuggingFaceEncoder(BaseEncoder):
```

Comment:
Can we make this top level rather than dynamically generated and explicitly pass in the required args?